File: app/src/main/java/ru/magflayer/spectrum/presentation/pages/splash/SplashActivity.java
package ru.magflayer.spectrum.presentation.pages.splash;

import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;

import com.arellomobile.mvp.presenter.InjectPresenter;

import ru.magflayer.spectrum.domain.injection.InjectorManager;
import ru.magflayer.spectrum.presentation.common.android.BaseActivity;
import ru.magflayer.spectrum.presentation.router.GlobalRouterImpl;

public class SplashActivity extends BaseActivity implements SplashView {

    private static final int CAMERA_PERMISSION_REQUEST = 111;

    @InjectPresenter
    SplashPresenter presenter;

    @Override
    protected void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        presenter.setRouter(new GlobalRouterImpl(this));

        int permissionCheck = ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA);
        if (permissionCheck != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this,
                    new String[]{Manifest.permission.CAMERA},
                    CAMERA_PERMISSION_REQUEST);
            return;
        }
        presenter.openMainPage();
    }

    @Override
    protected void inject() {
        InjectorManager.getAppComponent().inject(this);
    }

    @Override
    public void onRequestPermissionsResult(final int requestCode, @NonNull final String[] permissions,
                                           @NonNull final int[] grantResults) {
        switch (requestCode) {
            case CAMERA_PERMISSION_REQUEST: {
                if (grantResults.length > 0
                        && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    presenter.openMainPage();
                } else {
                    finish();
                }
                return;
            }
            default:
                super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        }
    }

    @Override
    public void closeScreen() {
        finish();
    }
}

File: app/src/main/java/ru/magflayer/spectrum/presentation/pages/splash/SplashActivity.kt
package ru.magflayer.spectrum.presentation.pages.splash

import android.Manifest
import android.content.pm.PackageManager
import android.os.Bundle
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import dagger.hilt.EntryPoint
import dagger.hilt.InstallIn
import dagger.hilt.android.AndroidEntryPoint
import dagger.hilt.android.EntryPointAccessors
import dagger.hilt.android.components.ActivityComponent
import moxy.presenter.InjectPresenter
import moxy.presenter.ProvidePresenter
import ru.magflayer.spectrum.R
import ru.magflayer.spectrum.presentation.common.android.BaseActivity
import ru.magflayer.spectrum.presentation.pages.main.MainActivity

@AndroidEntryPoint
class SplashActivity : BaseActivity(), SplashView {

    companion object {

        private const val CAMERA_PERMISSION_REQUEST = 111
        private val REQUIRED_PERMISSIONS = arrayOf(Manifest.permission.CAMERA)

    }

    @EntryPoint
    @InstallIn(ActivityComponent::class)
    interface SplashEntryPoint {
        fun splashPresenter(): SplashPresenter
    }

    @InjectPresenter
    lateinit var presenter: SplashPresenter

    @ProvidePresenter
    fun providePresenter(): SplashPresenter {
        return EntryPointAccessors.fromActivity(this, SplashEntryPoint::class.java)
            .splashPresenter()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (allPermissionsGranted()) {
            presenter.handlePermissionsGranted()
        } else {
            ActivityCompat.requestPermissions(
                this,
                REQUIRED_PERMISSIONS,
                CAMERA_PERMISSION_REQUEST
            )
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        when (requestCode) {
            CAMERA_PERMISSION_REQUEST -> if (allPermissionsGranted()) {
                presenter.handlePermissionsGranted()
            } else {
                Toast.makeText(this, R.string.camera_required, Toast.LENGTH_SHORT).show()
                finish()
            }
            else -> super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        }
    }

    override fun openMainScreen() {
        finish()
        startActivity(MainActivity.newIntent(this))
    }

    private fun allPermissionsGranted(): Boolean = REQUIRED_PERMISSIONS.all { permission ->
        ContextCompat.checkSelfPermission(this, permission) == PackageManager.PERMISSION_GRANTED
    }
}
